<div style="border:solid Chocolate 2px; padding: 40px">

<b> Мария, привет!👋</b>

Меня зовут Алексей Гриб, и я буду ревьюером твоего проекта. 

Сразу хочу предложить в дальнейшем общаться на "ты" - надеюсь, так будет комфортнее:) Но если это неудобно, обязательно дай знать, и мы придумаем что-нибудь ещё!
    
Цель ревью - не искать ошибки в твоём проекте, а помочь тебе сделать твою работу ещё лучше, устранив недочёты и приблизив её к реальным задачам специалиста по работе с данными. Поэтому не расстраивайся, если что-то не получилось с первого раза - это нормально, и это поможет тебе вырасти!
    
Ты можешь найти мои комментарии, обозначенные <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> похвала, рекомендации «со звёздочкой», полезные лайфхаки, которые сделают и без того красивое решение ещё более элегантным.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> некритичные ошибки или развивающие рекомендации на будущее. 
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
критичные ошибки, которые обязательно нужно исправить.
</div>

    
Пожалуйста, не удаляй мои комментарии, они будут особенно полезны для нашей работы в случае повторной проверки проекта. 
    
Ты также можешь задавать свои вопросы, реагировать на мои комментарии, делать пометки и пояснения - полная творческая свобода! Но маленькая просьба - пускай они будут отличаться от моих комментариев, это поможет избежать путаницы в нашем общении:)
Например, вот так:
    
<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

*твой текст*
</div>
    
Давай посмотрим на твой проект!

# Часть 1. Проверка гипотезы в Python и составление аналитической записки

Вы предобработали данные в SQL, и теперь они готовы для проверки гипотезы в Python. Загрузите данные пользователей из Москвы и Санкт-Петербурга c суммой часов их активности из файла yandex_knigi_data.csv. Если работаете локально, скачать файл можно по ссылке.

Проверьте наличие дубликатов в идентификаторах пользователей. Сравните размеры групп, их статистики и распределение.

Напомним, как выглядит гипотеза: пользователи из Санкт-Петербурга проводят в среднем больше времени за чтением и прослушиванием книг в приложении, чем пользователи из Москвы. Попробуйте статистически это доказать, используя одностороннюю проверку гипотезы с двумя выборками:

Нулевая гипотеза $H_0: \mu_{\text{СПб}} \leq \mu_{\text{Москва}}$ <br> Среднее время активности пользователей в Санкт-Петербурге не больше, чем в Москве.

Альтернативная гипотеза $H_1: \mu_{\text{СПб}} > \mu_{\text{Москва}}$ <br> Среднее время активности пользователей в Санкт-Петербурге больше, и это различие статистически значимо.

По результатам анализа данных подготовьте аналитическую записку, в которой опишите:

Выбранный тип t-теста и уровень статистической значимости.

Результат теста, или p-value.

Вывод на основе полученного p-value, то есть интерпретацию результатов.

Одну или две возможные причины, объясняющие полученные результаты.

## Напишите заголовок первой части проекта здесь

- Автор: Аверкина Мария
- Дата: 05.08.2025

## Цели и задачи проекта

Цель: Проверить следующую гипотезу: пользователи из Санкт-Петербурга проводят в среднем больше времени за чтением и прослушиванием книг в приложении, чем пользователи из Москвы.

Нулевая гипотеза $H_0: \mu_{\text{СПб}} \leq \mu_{\text{Москва}}$ <br> Среднее время активности пользователей в Санкт-Петербурге не больше, чем в Москве.

Альтернативная гипотеза $H_1: \mu_{\text{СПб}} > \mu_{\text{Москва}}$ <br> Среднее время активности пользователей в Санкт-Петербурге больше, и это различие статистически значимо.

Задачи:
1. Выгрузить данные из датасета. 
2. Проверить пользователей на дубликаты по их идентификаторам. 
3. Проверить размеры выборок: 

- Рассчитать количество уникальных пользователей из каждого города.

- Рассчитать и вывести на экран процентную разницу в количестве пользователей из разных городов. Построить любую удобную визуализацию, на которой будет видно возможное различие двух групп.

Для расчёта процентной разницы воспользуемся формулой:
$$P = 100 \cdot  \frac{|A − B|}{A}$$

4. Проверим, пересекаются ли пользователи. То есть, есть ли аномалия, при которой пользователь числится в обеих городах. 
5. Найдем распределения выборок из обеих городов (?)

## Описание данных

В датасете из файла `yandex_knigi_data.csv` содержатся следующие столбцы:

- `Unnamed: 0` - отдельный столбец с индексами;
- `city` - название города, в нашем случае Москва или Санкт-Петербург;
- `puid` - идентифекатор пользователя;
- `hours` - сумма длительности чтения или прослушивания в часах; 

## Содержимое проекта

Часть 1. Проверка гипотезы и составление аналитической записки. 
Часть 2. Анализ результатов А/В тестирования. 

---

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
<b>Все отлично!👍:</b> Хорошее вступление!
    
В нём есть всё, что необходимо, чтобы понять суть проекта с первых строк отчёта!

## 1. Загрузка данных и знакомство с ними

Загрузите данные пользователей из Москвы и Санкт-Петербурга c их активностью (суммой часов чтения и прослушивания) из файла `/datasets/yandex_knigi_data.csv`.

In [1]:
import pandas as pd

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Библиотеки импортировали - отлично!

In [2]:
books = pd.read_csv('/datasets/yandex_knigi_data.csv')

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> 
    
Данные загрузили.
    
При считывании данных из файла здорово перестраховывать себя от ошибок, связанных, например, с неверным указанием пути к файлу. А иногда бывает, что работаешь с файлом локально, выгружаешь его на сервер, ожидая, что он будет принимать данные, которые лежат на том же сервере, а код падает с ошибкой, потому что путь к файлу не поменялся с локального на серверный.
    
Для этого, например, можно использовать конструкцию `try-except`: сначала пробуешь локальный путь, при возникновении ошибки используется серверный путь (подробнее можешь почитать тут: https://pythonworld.ru/tipy-dannyx-v-python/isklyucheniya-v-python-konstrukciya-try-except-dlya-obrabotki-isklyuchenij.html).
    
Но еще лучше использовать библиотеку `os` - её использование позволит тебе проверять существование указанных директорий (что может быть актуально при одновременной работа на локальном и сетевом окружении) и загружать данные из существующей директории, избегая ошибок. Как пример:
    
    import os

    pth1 = '/folder_1/data.csv'
    pth2 = '/folder_2/data.csv'
    
    if os.path.exists(pth1):
        query_1 = pd.read_csv(pth1)
    elif os.path.exists(pth2):
        query_1 = pd.read_csv(pth2)
    else:
        print('Something is wrong')
    
Обрати внимание, что при считывании данных по `url` адресу (вроде `https://code.s3.yandex.net/datasets/`), `os` не сможет их проверить, так как работает только с физическими путями. В этом случае можно сделать проверку через `requests`: можно отправить `.get()` запрос и проверить `status_code` в ответе. Если он будет `200`, данные можно загружать.

Ещё на этапе считывания данных можно спарсить дату: за это действие отвечает параметр `parse_dates` метода `read_csv()`, в него нужно передать список с названием полей-дат, и в большинстве случаев дата будет корректно преобразована в нужный формат сразу:)
Также на этапе считывания данных можно задать индекс-столбец - за это действие отвечает параметр `index_col`.

In [3]:
books.head()

Unnamed: 0    city    puid       hours
0           0  Москва    9668   26.167776
1           1  Москва   16598   82.111217
2           2  Москва   80401    4.656906
3           3  Москва  140205    1.840556
4           4  Москва  248755  151.326434

In [4]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  8784 non-null   int64  
 1   city        8784 non-null   object 
 2   puid        8784 non-null   int64  
 3   hours       8784 non-null   float64
dtypes: float64(1), int64(2), object(1)
memory usage: 274.6+ KB


Промежуточный вывод:
- В датасете 8784 строки и 4 столбца, типы данных соотвествуют сути самих столбцов.

Далее проверим, встречаются ли дубликаты среди идентификаторов пользователей. 

In [5]:
# Счтиаем уникальное количество пользователей
n1 = books['puid'].nunique()

# Считаем общее кол-во пользователей
n2 = books['puid'].count()

# Считаем, сколько содержится дубликатов по столбцу
n3 = books.duplicated(subset='puid', keep='first').sum()

print(f'Уникальное количество пользователей {n1} отличается от общего колиечства пользователей {n2} на {n3}')

Уникальное количество пользователей 8540 отличается от общего колиечства пользователей 8784 на 244


- Очевидно, что встречаются дубликаты. Давайте их удалим.

In [6]:
# Удаляем дубликаты 
books = books.drop_duplicates(subset=['puid'])

# Проверяем, что дубликатов больше нет
display(books.duplicated(subset='puid', keep='first').sum())

0

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Удалены дубликаты по `puid`.

Посчитаем уникальное количесвто пользователей абсолютно и относительно для каждого города. Сравним разницу между ними и проверим, значимо ли это различие. 

In [7]:
# Найдем абсолютное уникальное количество пользователей в обеих городах
ma = books[books['city']=='Москва']['puid'].nunique()
spba = books[books['city']=='Санкт-Петербург']['puid'].nunique()

# Найдем относительные количества пользователей в обеих городах
mo = ma/n2
spbo = spba/n2

# Рассчитаем процентную разницу между городами
perc = round(100*((abs(ma-spba))/ma), 2)

print(f'Абсолютное количество пользователей в Москве составило {ma}, в Санкт-Петербурге {spba}. Относительное количество пользователей в Москве составило {mo}, в Санкт-Петербурге {spbo}. Процентная разница равна {perc} %.')

Абсолютное количество пользователей в Москве составило 6234, в Санкт-Петербурге 2306. Относительное количество пользователей в Москве составило 0.7096994535519126, в Санкт-Петербурге 0.26252276867030966. Процентная разница равна 63.01 %.


Можно заметить, что размеры выборок сильно отличаются (примерно на 63%). Это окажет сильное влияние на проведение дальнейших тестов. 

Проверим статистики каждого города по столбцу `hours`. 

In [8]:
group_1= books[books['city']=='Москва']['hours']
group_2 = books[books['city']=='Санкт-Петербург']['hours']

display(group_1.describe(),group_2.describe())

count    6234.000000
mean       10.881092
std        36.851683
min         0.000018
25%         0.059903
50%         0.924498
75%         5.939972
max       857.209373
Name: hours, dtype: float64

count    2306.000000
mean       11.264433
std        39.831755
min         0.000025
25%         0.060173
50%         0.875355
75%         6.138424
max       978.764775
Name: hours, dtype: float64

Несмотря на разницу в количестве пользователей, средние значения прослушанных и/или прочитанных часов равны, стандартные отклонения (корни из дисперсий) также приблизительно равны, минимальные и максимальные значения у обоих городов также схожи, сильных выбросов не наблюдается. В следующем пункте на основании полученных описаний статистик и размеров выборок применим тест Стьюдента. 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Есть сравнение групп по размеру выборок и статистикам.

## 2. Проверка гипотезы в Python

Гипотеза звучит так: пользователи из Санкт-Петербурга проводят в среднем больше времени за чтением и прослушиванием книг в приложении, чем пользователи из Москвы. Попробуйте статистически это доказать, используя одностороннюю проверку гипотезы с двумя выборками:

- Нулевая гипотеза H₀: Средняя активность пользователей в часах в двух группах (Москва и Санкт-Петербург) не различается.

- Альтернативная гипотеза H₁: Средняя активность пользователей в Санкт-Петербурге больше, и это различие статистически значимо.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Гипотезы сформулированы.

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> При работе с одосторонней проверкой нулевую гипотезу лучше формулировать от обратного к альтернативной: если альтернативная - `больше`, нулевая должна быть `не больше`.

In [9]:
from scipy.stats import ttest_ind

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b>
Все библиотеки стоит импортировать в начале проекта - таким образом получатель твоего отчёта сможет узнать о проблемах с его окружением сразу перед работой с проектом, а не посреди или в конце проекта.

In [10]:
group_2 = books[books['city']=='Москва']['hours']
group_1 = books[books['city']=='Санкт-Петербург']['hours']

test_result = ttest_ind (
    group_1, 
    group_2, 
    equal_var = False,
    alternative = 'greater'
)

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Мы проверяем гипотезу о том, что активность в Санкт-Петербурге выше, чем в Москвы - группа Санкт-Петербурга должна быть первой в тесте, а `alternative` должно быть `greater`.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Я изменила, но в чем смысл? Разве результат не такой же?
</div>

In [11]:
alpha = 0.05
pv = test_result.pvalue

print(f'test_result.pvalue = {pv}')

if test_result.pvalue<alpha:
    print('Нулевая гипотеза о том, что средняя активность пользователей в часах одинакова отвергается. Средняя активность пользователей в Санкт-Петербурге больше.')
else:
    print('Нулевую гипотезу о том, что средняя активность пользователей в часах одинакова отвергнуть не получилось.')

test_result.pvalue = 0.3435708118766314
Нулевую гипотезу о том, что средняя активность пользователей в часах одинакова отвергнуть не получилось.


Альтернативная гипотеза о том, что средняя активность пользователей в Санкт-Петербурге больше, и это различие статистически значимо не нашла подтверждения по результатам проведенного теста. 

## 3. Аналитическая записка
Аналитическая записка:

Был выбран тест Уэлча, так как размеры выборок достаточны для подобного теста, но при этом имеют сильные различия в размере, также дисперсии приблизительно равны, так как стандартные отклонения, являющиеся корнями из дисперсии, приблизительно равны

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> Несмотря на условие проекта использовать `ttest`, тут корректнее было бы применить тест Манна-Уитни из-за большого количества выбросов, которые не убираются даже по 99 персентилю.

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> C учётом большой разницы в размере выборок обязательно нужно применить поправку Уэлча.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Исправила, но вроде ничего не поменялось по результату.
</div>

Альтернативная гипотеза о том, что средняя активность пользователей в Санкт-Петербурге больше, и это различие статистически значимо не нашла подтверждения по результатам проведенного теста. P-value оказался больше установленного уровня значимости. То есть мы не смогли отвергнуть гипотезу о том, что средняя активность пользователей в часах одинакова.

Причины таких результатов: Изучая статистики в первом пункте, мы отметили, что все они почти равны между группами, нет ничего, что указавало бы на особую активность пользователей из Санкт-Петербурга. Их средние показатели близки к тем, что у пользователей из Москвы. Также размер группы пользователей из Санкт-Петербурга сильно меньше, чем из Москвы. Такой факт мог сыграть в пользу пользователей из Москвы и указать, что их средняя активность больше. Однако, гипотеза об одинаковой активности не отвергнута. 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Аналитическая записка по результатам тестирования сформирована, наполнение соответствует заданию.

----

# Часть 2. Анализ результатов A/B-тестирования

Теперь вам нужно проанализировать другие данные. Представьте, что к вам обратились представители интернет-магазина BitMotion Kit, в котором продаются геймифицированные товары для тех, кто ведёт здоровый образ жизни. У него есть своя целевая аудитория, даже появились хиты продаж: эспандер со счётчиком и напоминанием, так и подстольный велотренажёр с Bluetooth.

В будущем компания хочет расширить ассортимент товаров. Но перед этим нужно решить одну проблему. Интерфейс онлайн-магазина слишком сложен для пользователей — об этом говорят отзывы.

Чтобы привлечь новых клиентов и увеличить число продаж, владельцы магазина разработали новую версию сайта и протестировали его на части пользователей. По задумке, это решение доказуемо повысит количество пользователей, которые совершат покупку.

Ваша задача — провести оценку результатов A/B-теста. В вашем распоряжении:

* данные о действиях пользователей и распределении их на группы,

* техническое задание.

Оцените корректность проведения теста и проанализируйте его результаты.

## 1. Опишите цели исследования.



Цель: Оценка результатов A/B-теста.

Задачи: 

1. Загрузить данные оценить их целостность. 
2. Проверить пользователей, участвующих в тесте на соответсвие требованиям ТЗ, на равномерность распределения пользователей по группам теста и на отсутсвтие пересечений с конкурирующим тестом. 
3. Проанализировать данные о пользовательской активности, определить горизонт анализа, оценить достаточность выборки для получиения статистически значимых результатов A/B-теста. Рассчитать количество посетителей, сделавших покупку и общее кол-во для каждой группы. 
4. Проверить изменение конверсии подходящим статистическим тестом.

## 2. Загрузите данные, оцените их целостность.


In [12]:
participants = pd.read_csv('https://code.s3.yandex.net/datasets/ab_test_participants.csv')
events = pd.read_csv('https://code.s3.yandex.net/datasets/ab_test_events.zip',
                     parse_dates=['event_dt'], low_memory=False)

In [13]:
participants.head()

user_id group                  ab_test   device
0  0002CE61FF2C4011     B        interface_eu_test      Mac
1  001064FEAAB631A1     B  recommender_system_test  Android
2  001064FEAAB631A1     A        interface_eu_test  Android
3  0010A1C096941592     A  recommender_system_test  Android
4  001E72F50D1C48FA     A        interface_eu_test      Mac

In [14]:
events.head()

user_id            event_dt                        event_name  \
0            GLOBAL 2020-12-01 00:00:00  End of Black Friday Ads Campaign   
1  CCBE9E7E99F94A08 2020-12-01 00:00:11                      registration   
2            GLOBAL 2020-12-01 00:00:25                      product_page   
3  CCBE9E7E99F94A08 2020-12-01 00:00:33                             login   
4  CCBE9E7E99F94A08 2020-12-01 00:00:52                      product_page   

       details  
0  ZONE_CODE15  
1          0.0  
2          NaN  
3          NaN  
4          NaN

In [15]:
# Считаем сколько дубликатов в таблице
duplic_sum = events.duplicated(subset=['user_id', 'event_name'], keep='first').sum()

# Считаем сколько строк в таблице
events_len = len(events)

# Удаляем дубликаты
events = events.drop_duplicates(subset=['user_id', 'event_name'], keep='first')

# Смотрим сколько строк стало после удаления дубликатов
events_len_new = len(events)

print(f'В таблице events было найдено столько дубликатов по событиям {duplic_sum}, для наглядности отобразим сколько всего строк в таблице {events_len}. После удаления дубликатов строк в таблице стало - {events_len_new}')

В таблице events было найдено столько дубликатов по событиям 360249, для наглядности отобразим сколько всего строк в таблице 787286. После удаления дубликатов строк в таблице стало - 427037


<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Нужно проверить дубликаты по событиям.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Исправила, удалила пользователей, у которых были дубликаты в 'event_name'.
</div>

## 3. По таблице `ab_test_participants` оцените корректность проведения теста:

   3\.1 Выделите пользователей, участвующих в тесте, и проверьте:

   - соответствие требованиям технического задания,

   - равномерность распределения пользователей по группам теста,

   - отсутствие пересечений с конкурирующим тестом (нет пользователей, участвующих одновременно в двух тестовых группах).

Из выгрузки даннх стало понятно, что пользователи принимали участие не в одном тесте.

In [16]:
ab_test_participants = participants.groupby('ab_test')['user_id'].count()

display(ab_test_participants)

ab_test
interface_eu_test          10850
recommender_system_test     3675
Name: user_id, dtype: int64

Можно заметить, что в таблице существует информация о двух А/В тестах, давайте проверим, встречаются ли пользователи группы B, которые принимали участие в обоих тестах. Это может повлиять на независимость их мнения при проведении нашего теста. 

In [17]:
# Создаем отдельный датафрейм в котором посчитаем кол-во тестов у всех пользователей
participants_b = participants.groupby('user_id')['ab_test'].count().reset_index()
participants_b = participants_b.rename(columns={'ab_test': 'ab_test_count'})

participants = participants.merge(participants_b, how = 'left', on = 'user_id')

# Находим user_id, которые нужно удалить (кол-во тестов больше 1 и тестовая группа В)
users_to_remove = participants[
    (participants['ab_test_count'] > 1) & 
    (participants['ab_test'] == 'B')
]['user_id'].unique()

# Удаляем всех записей этих пользователей целиком
participants = participants[~participants['user_id'].isin(users_to_remove)]

# Удаляем временный столбец ab_test_count 
participants = participants.drop(columns=['ab_test_count'])

display(participants)

user_id group                  ab_test   device
0      0002CE61FF2C4011     B        interface_eu_test      Mac
1      001064FEAAB631A1     B  recommender_system_test  Android
2      001064FEAAB631A1     A        interface_eu_test  Android
3      0010A1C096941592     A  recommender_system_test  Android
4      001E72F50D1C48FA     A        interface_eu_test      Mac
...                 ...   ...                      ...      ...
14520  FFE7FC140521F5F6     A        interface_eu_test       PC
14521  FFEFC0E55C1CCD4F     A        interface_eu_test       PC
14522  FFF28D02B1EACBE1     B  recommender_system_test       PC
14523  FFF28D02B1EACBE1     B        interface_eu_test       PC
14524  FFF58BC33966EB51     A        interface_eu_test  Android

[14525 rows x 4 columns]

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Нас интересует, есть ли пользователи, попавшие в группу В в двух тестах сразу. Если такие пользователи будут, по ним мы не сможем оценить эффект от нововведения, так как не сможем понять, под воздействием какого нововведения было совершено целевое событие - от изменений в нашем тесте или в конкурирующем. Таких пользователей лучше всего удалить для корректной оценки метрик нашего теста.
    
Тут нужно работать с исходой таблицей `participants`, предварительно отфильтровав пользователей группы В, и если такие пользователи будут, то удалять их нужно из `participants` со срезом по тесту `interface_eu_test`, который нужно выделить в начале этого раздела согласно условию проекта:
    
    Выделите пользователей, участвующих в тесте, и проверьте:

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Исправила.
</div>

Проверим равномерность распределения пользователей по группам А/В теста. 

In [18]:
participants = participants[participants['ab_test']== 'interface_eu_test']

# Считаем, сколько участников в группе А и В интересущего нас теста
group_a = participants[participants['group']=='A']['user_id'].nunique()
group_b = participants[participants['group']=='B']['user_id'].nunique()

# Считаем переменную с общим кол-вом пользователей интересущего нас теста
total_u = participants[participants['ab_test']== 'interface_eu_test']['user_id'].nunique()

# Считаем относительные значения кол-ва участников от общего кол-ва по обеим группам
dev_a = group_a/total_u*100
dev_b = group_b/total_u*100

# Сичтаем относительную разницу по группам по модулю
dev_ab = abs(dev_a-dev_b)

print(f'Всего участников в группе А - {group_a}, в группе В - {group_b}.')

print(f'В группе А относительное значение участников от общего кол-ва составило - {dev_a}%, в группу В - {dev_b}%, разница между группами в количестве участников - {dev_ab}%')

Всего участников в группе А - 5383, в группе В - 5467.
В группе А относительное значение участников от общего кол-ва составило - 49.612903225806456%, в группу В - 50.38709677419355%, разница между группами в количестве участников - 0.7741935483870961%


Участники рапсределены равномерно, разница в количестве между группами составила менее 1%. 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Распределение пользователей по группам равномерное, верно.

Проверим, нет ли пересечений пользователей по группам, то есть нет такой ситуации, что пользователей попал и в группу А, и в группу В.

In [19]:
grp = participants.groupby('user_id')['group'].nunique().reset_index()

grp_ab = grp[grp['group']>1]['user_id'].count()

display(grp_ab)

0

Пересечений не найдено. 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Внутри исследуемого теста нет пересечений пользователей между группами.

3\.2 Проанализируйте данные о пользовательской активности по таблице `ab_test_events`:

- оставьте только события, связанные с участвующими в изучаемом тесте пользователями;

In [20]:
# Объединяем две таблицы для дальнейшей фильтрации
participants_events = participants.merge(events, on='user_id', how='inner')

# Фильтруем данные новой таблицы только для интересущего нас теста 
participants_events = participants_events[participants_events['ab_test']== 'interface_eu_test']

display(participants_events)

user_id group            ab_test   device            event_dt  \
0      0002CE61FF2C4011     B  interface_eu_test      Mac 2020-12-07 04:37:31   
1      0002CE61FF2C4011     B  interface_eu_test      Mac 2020-12-07 04:37:49   
2      0002CE61FF2C4011     B  interface_eu_test      Mac 2020-12-08 22:15:36   
3      0002CE61FF2C4011     B  interface_eu_test      Mac 2020-12-10 15:23:09   
4      001064FEAAB631A1     A  interface_eu_test  Android 2020-12-20 14:12:45   
...                 ...   ...                ...      ...                 ...   
36125  FFF28D02B1EACBE1     B  interface_eu_test       PC 2020-12-16 08:25:31   
36126  FFF28D02B1EACBE1     B  interface_eu_test       PC 2020-12-21 15:22:57   
36127  FFF28D02B1EACBE1     B  interface_eu_test       PC 2020-12-24 00:54:54   
36128  FFF58BC33966EB51     A  interface_eu_test  Android 2020-12-13 17:37:34   
36129  FFF58BC33966EB51     A  interface_eu_test  Android 2020-12-13 17:56:02   

         event_name details  
0      registration   -2.38  
1             login     NaN  
2      product_page     NaN  
3      product_cart     NaN  
4      registration   -2.14  
...             ...     ...  
36125         login     NaN  
36126  product_page     NaN  
36127  product_cart     NaN  
36128  registration   -0.45  
36129         login     NaN  

[36130 rows x 7 columns]

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Отобраны события, относящиеся только к нашему тесту.

Далее будем работать с таблицей `participants_events`.

- определите горизонт анализа: рассчитайте время (лайфтайм) совершения события пользователем после регистрации и оставьте только те события, которые были выполнены в течение первых семи дней с момента регистрации;

In [21]:
# Группируем данные по двум необходимым столбцам с идентификатором пользователя и названием мероприятия, оставляя дату
participants_events_all = participants_events.groupby(['user_id', 'event_name', 'group'], as_index=False)['event_dt'].first()

In [22]:
# Создаем табличку, где будут идентификаторы пользователей и даты для мероприятия, которое назваетя регистрация
# переименуем столбец с датой
participants_events_reg = participants_events_all[participants_events['event_name'] == 'registration'][['user_id','event_dt', 'group']]
participants_events_reg = participants_events_reg.rename(columns={'event_dt': 'reg_dt'})

# Создадим похожую на первую таблицу, только в нее войдут все меропрития, которые НЕ назваются регистрация
participants_events_no_reg = participants_events_all[participants_events_all['event_name'] != 'registration'][['user_id','event_dt', 'event_name']]
participants_events_no_reg = participants_events_no_reg.rename(columns={'event_name': 'event_name_without_reg'})

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Из основной таблицы не нужно удалять событие с регистрацией - в противном случае мы не сможем посчитать всех пользователей, принявших участие в тесте.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Из основной таблицы я ничего не удаляла, она у нас называется participants_events, в ней объединены изначальные датафреймы participants и events, с фильтрацией по интересующему нас тесту, удаленными дубликаты и остальной необходимой предобработкой. В этом задании я создаю отдельные таблички и на их основании вывожу лайфтайм, не совсем понимаю, в чем тут проблема. Вроде я выполнила само задание и лайфтайм можно посмотреть, если необходимо будет посчитать всех пользователей, принявших участие в тесте, это можно будет сделать на оснвоании датафрема participants_events.
</div>

In [23]:
# Объединим две таблички
pne_pne = participants_events_no_reg.merge(participants_events_reg, how = 'left', on = 'user_id')

# Создадим столбец с разнице между датами
pne_pne['diff_dt'] = pne_pne['event_dt'] - pne_pne['reg_dt']

# Отфильтруем нашу табличку
pne_pne = pne_pne[pne_pne['diff_dt'] <= '7 days']

display(pne_pne)

user_id            event_dt event_name_without_reg  \
0      0002CE61FF2C4011 2020-12-07 04:37:49                  login   
1      0002CE61FF2C4011 2020-12-10 15:23:09           product_cart   
2      0002CE61FF2C4011 2020-12-08 22:15:36           product_page   
3      001064FEAAB631A1 2020-12-20 14:43:27                  login   
4      001064FEAAB631A1 2020-12-21 03:20:38           product_page   
...                 ...                 ...                    ...   
25274  FFE7FC140521F5F6 2020-12-26 14:37:21               purchase   
25275  FFEFC0E55C1CCD4F 2020-12-13 23:52:22                  login   
25276  FFF28D02B1EACBE1 2020-12-16 08:25:31                  login   
25278  FFF28D02B1EACBE1 2020-12-21 15:22:57           product_page   
25279  FFF58BC33966EB51 2020-12-13 17:56:02                  login   

                   reg_dt group         diff_dt  
0     2020-12-07 04:37:49     B 0 days 00:00:00  
1     2020-12-07 04:37:49     B 3 days 10:45:20  
2     2020-12-07 04:37:49     B 1 days 17:37:47  
3     2020-12-20 14:43:27     A 0 days 00:00:00  
4     2020-12-20 14:43:27     A 0 days 12:37:11  
...                   ...   ...             ...  
25274 2020-12-23 09:21:10     A 3 days 05:16:11  
25275 2020-12-13 23:52:22     A 0 days 00:00:00  
25276 2020-12-16 08:25:31     B 0 days 00:00:00  
25278 2020-12-16 08:25:31     B 5 days 06:57:26  
25279 2020-12-13 17:56:02     A 0 days 00:00:00  

[23897 rows x 6 columns]

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Фильтрация событий по `lifetime` выполнена.

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Фильтровать нужно до седьмого дня невключительно, так как нужны дни с 0 по 6. Обрати также внимание на столбцы с постфиксами `_x` и `_y` - их наличие указывает на неоптимальный способ соединения данных. Стоит переделать так, чтобы дубликатов с постфиксами не было.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Исправила.
</div>

Оцените достаточность выборки для получения статистически значимых результатов A/B-теста. Заданные параметры:

- базовый показатель конверсии — 30%,

- мощность теста — 80%,

- достоверность теста — 95%.

In [24]:
from statsmodels.stats.power import NormalIndPower
from statsmodels.stats.proportion import proportion_effectsize

# Задаём параметры
alpha = 0.05  # Уровень значимости
beta = 0.2 # Ошибка второго рода, часто 1 - мощность
power = 0.8  # Мощность теста
p = 0.3 # Базовый уровень доли
mde = 0.03  # Минимальный детектируемый эффект

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> `mde` составляет `0.03` - умножать на `p` не нужно.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Спасибо большое за ответ! Исправила.
</div>

In [25]:
effect_size = proportion_effectsize(p, p + mde)

In [26]:
# Инициализируем класс NormalIndPower
power_analysis = NormalIndPower()

# Рассчитываем размер выборки
sample_size = power_analysis.solve_power(
    effect_size = effect_size,
    power = power,
    alpha = alpha,
    ratio = 1 # Равномерное распределение выборок
)

print(f"Необходимый размер выборки для каждой группы: {int(sample_size)}")

Необходимый размер выборки для каждой группы: 3761


<div class="alert alert-info"> <b>Комментарий студента:</b> Я не понимаю, как иначе тут оценить достаточность выборки. Если это неправильно, пожалуйста, напишите в каком направлении двигатья, у меня ноль идей. </div>


<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Выше прокомментировал, основная ошибка - неверный расчёт `mde`. Нужно сделать вывод по результатам сравнения.

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> Обрати внимание, что и онлайн-калькутяор Эвана Миллера, и `NormalIndPower` с `alternative='two-sided'` проверяет размер выборки для __двусторонней__ гипотезы - мы же работаем с __односторонней__, поэтому расчётное значение выборки выше, чем нужно для одностороннего теста.

- рассчитайте для каждой группы количество посетителей, сделавших покупку, и общее количество посетителей.

In [27]:
participants_events['event_name'].unique()

pne_purch_a = participants_events[(participants_events['group']=='A')&(participants_events['event_name']=='purchase')]['user_id'].nunique()
pne_a_tot = participants_events[participants_events['group']=='A']['user_id'].nunique()

part_purch_a = round(pne_purch_a/pne_a_tot, 2)*100

pne_purch_b = participants_events[(participants_events['group']=='B')&(participants_events['event_name']=='purchase')]['user_id'].nunique()
pne_b_tot = participants_events[participants_events['group']=='B']['user_id'].nunique()

part_purch_b = round(pne_purch_b/pne_b_tot,2)*100

print(f'Всего пользователей в группе А = {pne_a_tot}, кол-вопользователей в группе А, сделваших покупку = {pne_purch_a}. Всего пользователей в группе В = {pne_b_tot}, , кол-вопользователей в группе В, сделваших покупку =  {pne_purch_b}. Отношение пользователей сделавших покупку к общему числу пользователей в группе А = {part_purch_a} %. Отношение пользователей сделавших покупку к общему числу пользователей в группе B = {part_purch_b} %.')

Всего пользователей в группе А = 5383, кол-вопользователей в группе А, сделваших покупку = 1766. Всего пользователей в группе В = 5467, , кол-вопользователей в группе В, сделваших покупку =  1931. Отношение пользователей сделавших покупку к общему числу пользователей в группе А = 33.0 %. Отношение пользователей сделавших покупку к общему числу пользователей в группе B = 35.0 %.


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Пользователи и конверсия по группам посчитаны.

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> Удобнее было бы выполнить это задание с помощью сводной таблицы:
        
    df[df['event_name'].isin(['registration', 'purchase'])].pivot_table(index='group',
                                                                        columns='event_name',
                                                                        values='user_id',
                                                                        aggfunc='nunique')

В полученной таблице существенно проще посчитать конверсию.

- сделайте предварительный общий вывод об изменении пользовательской активности в тестовой группе по сравнению с контрольной.

Если судить по рассчету пользователей, которые делали покупки, то можно наблюдать небольшой прирост в группе В. Более точные выводы можно будет сделать в конце, после оценки результатов А/B тестирования. 

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Выводы нужно будет актуализировать, так как конверсии посчитаны неправильно из-за ошибок на предыдущих шагах. Акцент нужно сделать на сравнении фактической и плановой конверсии - был ли достигнут плановый прирост.

<div class="alert alert-info"> <b>Комментарий студента:</b> Исправила. </div>

## 4. Проведите оценку результатов A/B-тестирования:

- Проверьте изменение конверсии подходящим статистическим тестом, учитывая все этапы проверки гипотез.

Гипотеза заключается в следующем: упрощение интерфейса приведёт к тому, что в течение семи дней после регистрации в системе конверсия зарегистрированных пользователей в покупателей увеличится как минимум на три процентных пункта.

Нулевая гипотеза (Н0):  В течение семи дней после регистрации в системе конверсия зарегистрированных пользователей в покупателей не изменилась. 

Альтернативная гипотеза (Н1): В течение семи дней после регистрации в системе конверсия зарегистрированных пользователей в покупателей увеличилась как минимум на 3%.

In [28]:
from statsmodels.stats.proportion import proportions_ztest

# Размеры выборок А и В
n_a = participants_events[participants_events['group']=='A']['user_id'].nunique() 
n_b = participants_events[participants_events['group']=='B']['user_id'].nunique()


# Кол-во успехов (покупок) в группе А
m_a = participants_events[(participants_events['event_name'] == 'purchase')&(participants_events['group']=='A')]['user_id'].nunique()

# Кол-во успехов (покупок) в группе В
m_b = participants_events[(participants_events['event_name'] == 'purchase')&(participants_events['group']=='B')]['user_id'].nunique()

p_a = m_a/n_a

p_b = m_b/n_b

print(f'n_a = {n_a}, n_b = {n_b}')
print(f'm_a = {m_a}, m_b = {m_b}')
print(f'p_a = {p_a}, p_b = {p_b}')

alpha = 0.05

stat_ztest, p_value_ztest = proportions_ztest(
    [m_a, m_b], 
    [n_a, n_b], 
    alternative = 'smaller'
    )

print(f'p_value = {p_value_ztest}')

if p_value_ztest<alpha:
    print('Нулевая гипотеза о том, что в течение семи дней после регистрации в системе конверсия зарегистрированных пользователей в покупателей не изменилась отвергается. В течение семи дней после регистрации в системе конверсия зарегистрированных пользователей изменилась минимум на 3 %.')
else:
    print('Нулевую гипотезу о том, что в течение семи дней после регистрации в системе конверсия зарегистрированных пользователей в покупателей не изменилась отвергнуть не получилось.')

n_a = 5383, n_b = 5467
m_a = 1766, m_b = 1931
p_a = 0.3280698495262865, p_b = 0.35321017011157857
p_value = 0.0028679563964638466
Нулевая гипотеза о том, что в течение семи дней после регистрации в системе конверсия зарегистрированных пользователей в покупателей не изменилась отвергается. В течение семи дней после регистрации в системе конверсия зарегистрированных пользователей изменилась минимум на 3 %.


<div class="alert alert-info"> <b>Комментарий студента:</b> Я отчаялась. Я не понимаю, какой тест тут применить. Z-тест, потому что конверсия это пропорция?? Но по описанию Z-тест здесь не подходит. Тест Стьюдента?? Но как посчитать метрики. Я решила посчитать просто конверсию как кол-во пользователей, сделавших покупку/общему числу пользователей на этапе регистрации (по группам). Пусть будет хоть это. Пожалуйста, направьте, какой тетст тут использовать, как метрики нормально посчитать, я не справляюсь. </div>

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Здесь нужен `proportion_ztest`, так как мы сравниваем конверсии двух групп.

<div class="alert alert-info"> <b>Комментарий студента:</b> Спасибо, исправила. </div>

- Опишите выводы по проведённой оценке результатов A/B-тестирования. Что можно сказать про результаты A/B-тестирования? Был ли достигнут ожидаемый эффект в изменении конверсии?

Да, ожидаемый эффект был достигнут. Конверсия в экспериментальной группе В изменилась, это означает, что новый, упрощенный интерфейс правда помог пользователям лучше ориентироваться на сайте.

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Итоговый вывод по завершении работы над проектом нужно будет дописать.

<div class="alert alert-info"> <b>Комментарий студента:</b> Дописала. </div>

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ревьюера: общий вывод по проекту.

Мария, проект получился на довольно хорошем уровне - отличная работа над проектом, молодец!

Мне нравится твой аналитический подход к выполнению проекта, ты соблюдаешь структуру работы, выполняешь её последовательно - это очень хорошо! Шаги проекта выполнены по порядку согласно плану проекта, нет смысловых и структурных ям. Важно, что не забываешь про выводы.
    
Над проектом ещё стоит поработать - есть рекомендации по дополнению некоторых твоих шагов проекта. Такие рекомендации я отметил жёлтыми комментариями. Будет здорово, если ты учтёшь их - так проект станет структурно и содержательно более совершенным.
    
Также в работе есть критические замечания. К этим замечаниям я оставил пояснительные комментарии красного цвета, в которых перечислил возможные варианты дальнейших действий. Уверен, ты быстро с этим управишься:)
    
Если о том, что нужно сделать в рамках комментариев, будут возникать вопросы - оставь их, пожалуйста, в комментариях, и я отвечу на них во время следующего ревью.
    
Также буду рад ответить на любые твои вопросы по проекту или на какие-либо другие, если они у тебя имеются - оставь их в комментариях, и я постараюсь ответить:)
    
Жду твой проект на повторном ревью. До встречи:)